### PR TITLE
Refactor Carousel to use Hooks

### PR DESCRIPTION
**Pull request by sivakar12**
#### What does this PR do?
Refactors the Carousel component to use Hooks
#### Where should the reviewer start?
src/js/components/Carousel.js
#### What testing has been done on this PR?
Ran the jest tests.
The snapshots had to be updated.
#### How should this be manually tested?
By viewing storybook stories
#### Any background context you want to provide?
I made a small change to the logic. Previously, boundaries are checked and onLeft and onRight functions are added to DOM elements conditionally. Now, the boundary conditions are checked inside the onLeft and onRight callbacks. That caused the snapshots to be updated.
I thought this might be cleaner. If you think otherwise, I can change it back.
#### What are the relevant issues?
#3394 
#### Do the grommet docs need to be updated?
No
#### Should this PR be mentioned in the release notes?
No
#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

### CODE REVIEW

File: src/js/components/Carousel/__tests__/__snapshots__/Carousel-test.js.snap [hunk 1]
```diff
@@ -323,6 +323,7 @@ exports[`Carousel basic 1`] = `
           className="c8"
           disabled={true}
           onBlur={[Function]}
+          onClick={[Function]}
```
Comment (ShimiSun) [inline src/js/components/Carousel/__tests__/__snapshots__/Carousel-test.js.snap]: Usually, when snapshots changed on refactoring it means that the functionality might have changed, will you be willing to take a look at why this snapshot changed? might mean we are doing something different with `onClick()`

Comment (sivakar12) [reply src/js/components/Carousel/__tests__/__snapshots__/Carousel-test.js.snap]: I explained it in the background context

Comment (ShimiSun) [reply src/js/components/Carousel/__tests__/__snapshots__/Carousel-test.js.snap]: thanks, I'll re-read

**Comment by sivakar12** [comment]
I noticed that all the functional components that are wrapped with `withTheme` and `withFocus` also have `withForwardRef`.
Should I wrap this with `withForwardRef` as well?

**Comment by ericsoderberghp** [comment]
> I noticed that all the functional components that are wrapped with `withTheme` and `withFocus` also have `withForwardRef`.
> Should I wrap this with `withForwardRef` as well?

We're slowly moving away from higher order components and recompose. With hooks, we're using `useContext()` for theme, and `forwardRef()` for refs. You can see an example of this in Button.js https://github.com/grommet/grommet/blob/master/src/js/components/Button/Button.js

**Comment by sivakar12** [comment]
> > I noticed that all the functional components that are wrapped with `withTheme` and `withFocus` also have `withForwardRef`.
> > Should I wrap this with `withForwardRef` as well?
> 
> We're slowly moving away from higher order components and recompose. With hooks, we're using `useContext()` for theme, and `forwardRef()` for refs. You can see an example of this in Button.js https://github.com/grommet/grommet/blob/master/src/js/components/Button/Button.js

OK.
I want to know whether this particular component needs to forward the ref? I am getting some warning when I run the tests? The previous class component was not doing anything with the refs, however

**Comment by ericsoderberghp** [comment]
If there were no refs before, no need to add them just to convert to hooks. ref can be added later.

**Comment by sivakar12** [comment]
```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
      
      Check the render method of `WithTheme(Carousel)`.
          in Carousel (created by WithTheme(Carousel))
          in WithTheme(Carousel) (created by FocusableComponent)
          in FocusableComponent (created by WithTheme(Carousel))
          in WithTheme(Carousel)
          in div (created by StyledGrommet)
          in StyledGrommet (created by Grommet)
          in Grommet
```

This is the warning I get. withFocus or withTheme is assigning a ref to this component. When it was a class component it was fine. But functional components can't have one.
Shall I leave it like this since this is only a warning and our tests are all passing? Or shall I take the forwardRef as give it to some child (the Stack?)?

**Comment by ericsoderberghp** [comment]
My apologies. I missed the detail that `recompose` was still being used. We should be removing `recompose`, `withTheme`, and `withFocus`. You can take a look at Button.js to see how this is done. Basically, replace `withTheme` with `const theme = useContext(ThemeContext);` and replace `withFocus` with `const [focus, setFocus] = useState();` along with `onFocus` and `onBlur` handlers.

File: src/js/components/Carousel/Carousel.js [hunk 2]
```diff
@@ -204,9 +224,6 @@ if (process.env.NODE_ENV !== 'production') {
   // eslint-disable-next-line global-require
   CarouselDoc = require('./doc').doc(Carousel);
 }
-const CarouselWrapper = compose(
-  withFocus(),
-  withTheme,
-)(CarouselDoc || Carousel);
+const CarouselWrapper = CarouselDoc || Carousel;
```
Comment (ericsoderberghp) [inline src/js/components/Carousel/Carousel.js]: Also need `Carousel.displayName = 'Carousel';` around line 221.

**Comment by ericsoderberghp** [comment]
Thanks for your contribution!

**Pull request merged**